### PR TITLE
[AudioToolbox] Update bindings to Xcode 9 beta2.

### PR DESCRIPTION
**Pull request by mandel-macaque**

### CODE REVIEW

File: src/AudioUnit/AudioComponent.cs [hunk 1]
```diff
@@ -224,6 +224,29 @@ public XamCore.AppKit.NSImage GetIcon ()
...
+		[NoWatch, NoTV, Mac (10,13), iOS (11,0)]
+		[DllImport (Constants.AudioUnitLibrary, EntryPoint = "AudioUnitExtensionSetComponentList")]
+		static extern int /* OSStatus */ SetComponentList (IntPtr /* CFString */ extensionIdentifier, /* CFArrayRef */ IntPtr audioComponentInfo);
+
+		[NoWatch, NoTV, Mac (10,13), iOS (11,0)]
+		[DllImport (Constants.AudioUnitLibrary, EntryPoint = "AudioUnitExtensionCopyComponentList")]
+		static extern /* CFArrayRef */ IntPtr CopyComponentList (IntPtr /* CFString */ extensionIdentifier);
+
+		[NoWatch, NoTV, Mac (10,13), iOS (11,0)]
+		public NSArray ComponentList {
```
Comment (rolfbjarne) [inline src/AudioUnit/AudioComponent.cs]: Manual binding code needs manual tests :)

File: src/AudioUnit/AudioUnit.cs [hunk 2]
```diff
@@ -352,6 +356,59 @@ public enum AudioUnitClumpID // UInt32 in AudioUnitParameterInfo
 		System = 0
 	}
 
+	public enum AUParameterEventType : uint
+	{
+		Immediate = 1,
+		Ramped = 2
```
Comment (rolfbjarne) [inline src/AudioUnit/AudioUnit.cs]: Minor: comma at the end of the last entry too, to minimize future diffs.

File: src/AudioUnit/AudioUnit.cs [hunk 3]
```diff
@@ -873,6 +930,11 @@ public AudioUnitStatus SetParameter (AudioUnitParameterType type, float value, A
 			return AudioUnitSetParameter (handle, type, scope, audioUnitElement, value, 0);
 		}
 		
+		public AudioUnitStatus ScheduleParameter (AudioUnitParameterType type, AudioUnitParameterEvent inParameterEvent, uint inNumParamEvents)
+		{
+			return 0;
```
Comment (rolfbjarne) [inline src/AudioUnit/AudioUnit.cs]: This doesn't look complete :smile:

File: src/AudioUnit/AudioUnit.cs [hunk 4]
```diff
@@ -352,6 +356,59 @@ public enum AudioUnitClumpID // UInt32 in AudioUnitParameterInfo
...
+	{
+		Immediate = 1,
+		Ramped = 2
+	}
+
+	[StructLayout (LayoutKind.Sequential)]
+	public struct AudioUnitParameterEvent
+	{
+		public uint scope;
+
```
Comment (rolfbjarne) [inline src/AudioUnit/AudioUnit.cs]: No need for whitespace between fields.

File: src/AudioUnit/AudioUnit.cs (hunk 3 above)
Comment (mandel-macaque) [inline src/AudioUnit/AudioUnit.cs]: Ups! I missed that, mea culpa

File: src/AudioUnit/AudioComponent.cs (hunk 1 above)
Comment (mandel-macaque) [inline src/AudioUnit/AudioComponent.cs]: ACK

File: src/AudioUnit/AudioComponent.cs [hunk 5]
```diff
@@ -224,6 +224,29 @@ public XamCore.AppKit.NSImage GetIcon ()
...
+			get {
+				using (var cfString = new CFString (Name)) {
+					return new NSArray (CopyComponentList (cfString.Handle));
```
Comment (dalexsoto) [inline src/AudioUnit/AudioComponent.cs]: I think you want `Runtime.GetNSObject<NSArray>(CopyComponentList (cfString.Handle));` instead or if we know the type inside the array we can use any of the NSArray.Fromxx to get a strongly typed array

File: src/AudioUnit/AudioUnit.cs [hunk 6]
```diff
@@ -1954,7 +2011,8 @@ public enum AUSpatializationAlgorithm : uint
 		Hrtf = 2,
 		SoundField = 3,
 		VectorBasedPanning = 4,
-		StereoPassThrough = 5
+		StereoPassThrough = 5,
+		HRTFHQ = 6,
```
Comment (spouliot) [inline src/AudioUnit/AudioUnit.cs]: casing ^

File: src/audiounit.cs [hunk 7]
```diff
@@ -39,6 +39,8 @@ namespace XamCore.AudioUnit {
 			uint frameCount, nint inputBusNumber, AudioBuffers inputData);
 
 	delegate void AUScheduleParameterBlock (AUEventSampleTime eventSampleTime, uint rampDurationSampleFrames, ulong parameterAddress, float value);
+	[iOS (11, 0), Mac (10,13), TV (11,0), NoWatch]
+	delegate int AUMidiOutputEventBlock (long eventSampleTime, byte cable, nint length, ref byte midiBytes);
```
Comment (spouliot) [inline src/audiounit.cs]: incorrect,
`typedef OSStatus (^AUMIDIOutputEventBlock)(AUEventSampleTime eventSampleTime, uint8_t cable, NSInteger length, const uint8_t *midiBytes);`

The last argument is *not* a single byte, but a pointer to an array of bytes.

File: src/audiounit.cs [hunk 8]
```diff
@@ -238,6 +261,14 @@ interface AUAudioUnit
 	[BaseType (typeof (AUAudioUnit))]
 	interface AUAudioUnit_AUAudioInputOutputUnit {
 
+		[Mac (10,12), NoTV, NoiOS, NoWatch]
+		[Export ("deviceID")]
+		uint GetDeviceID ();
```
Comment (spouliot) [inline src/audiounit.cs]: `Id` <- not an acronym

File: src/audiounit.cs [hunk 9]
```diff
@@ -238,6 +261,14 @@ interface AUAudioUnit
...
+
+		[Mac (10,12), NoTV, NoiOS, NoWatch]
+		[Export ("setDeviceID:error:")]
+		bool SetDeviceID (uint deviceID, out NSError outError);
```
Comment (spouliot) [inline src/audiounit.cs]: `Id` <- not an acronym

File: src/audiounit.cs [hunk 10]
```diff
@@ -273,7 +304,20 @@ interface AUAudioUnit_AUAudioInputOutputUnit {
...
+		[Export ("deviceInputLatency")]
+		double GetDeviceInputLatency ();
+
+		[Mac (10, 13), NoWatch, NoiOS, NoTV]
+		[Export ("deviceOutputLatency")]
+		double GetDeviceOutputLatency ();
+
+		[Watch (4, 0), TV (11, 0), Mac (10, 13), iOS (11, 0)]
+		[Export ("running")]
+		bool IsRunning ();
```
Comment (spouliot) [inline src/audiounit.cs]: maybe a property ? how's swift binding it ?

File: src/AudioUnit/AudioComponent.cs [hunk 11]
```diff
@@ -38,6 +38,159 @@
...
+		static NSString userClientK = new NSString ("iokit.user-client");
+		static NSString globalNameK = new NSString ("mach-lookup.global-name");
+		static NSString networkClientK = new NSString ("network.client");
+		static NSString exceptionK = new NSString ("temporary-exception.files.all.read-write");
+
+		public ResourceUsageInfo () : base () {}
+
+		public ResourceUsageInfo (NSDictionary dic) : base (dic) {}
+
+		public NSString[] IOKitUserClient {
```
Comment (spouliot) [inline src/AudioUnit/AudioComponent.cs]: why not `string[]` ?

File: src/AudioUnit/AudioComponent.cs [hunk 12]
```diff
@@ -38,6 +38,159 @@
...
+		public NSString[] IOKitUserClient { 
+			get {
+				return GetArray<NSString> (userClientK);
+			} 
+			set {
+				SetArrayValue <NSString> (userClientK, value);
+			}
+		}
+
+		public NSString[] MachLookUpGlobalName {
```
Comment (spouliot) [inline src/AudioUnit/AudioComponent.cs]: why not `string[]` ?

File: src/audiounit.cs (hunk 10 above)
Comment (mandel-macaque) [inline src/audiounit.cs]: AUAudioUnit_AUAudioInputOutputUnit is a protocol, AFAIK we cannot have optional properties and we have to have methods. Same thing happens with all the other methods GetDeviceOutputLatency, GetDeviceInputLatency, GetOutputProvider etc...

File: src/AudioUnit/AudioUnit.cs (hunk 6 above)
Comment (mandel-macaque) [inline src/AudioUnit/AudioUnit.cs]: I think this is an acronym for Head-related-transfer-function and no idea about the HQ.. but will lower everything except the H.

**Comment by mandel-macaque** [comment]
Failures seem not to be related to the changes in audio tools.

File: src/AudioUnit/AudioComponent.cs [hunk 13]
```diff
@@ -224,6 +377,58 @@ public XamCore.AppKit.NSImage GetIcon ()
...
+		static extern /* CFArrayRef */ IntPtr CopyComponentList (IntPtr /* CFString */ extensionIdentifier);
+
+		[NoWatch, NoTV, Mac (10,13), iOS (11,0)]
+		public AudioComponentInfo[] ComponentList {
+			get {
+				using (var cfString = new CFString (Name)) {
+					var cHandle = CopyComponentList (cfString.Handle);
+					if (cHandle == IntPtr.Zero)
+						return null;
+					using (var nsArray = Runtime.GetNSObject<NSArray> (cHandle)) {
```
Comment (rolfbjarne) [inline src/AudioUnit/AudioComponent.cs]: You own the returned CFArray, so use `GetNSObject<NSArray> (cHandle, owns: true)`.

File: src/AudioUnit/AudioComponent.cs [hunk 14]
```diff
@@ -224,6 +377,58 @@ public XamCore.AppKit.NSImage GetIcon ()
...
+						var result = (AudioConverterError) SetComponentList (cfString.Handle, array.Handle);
+						switch (result) {
+						case AudioConverterError.HardwareInUse:
+							throw new InvalidOperationException ("ComponentList could not be set: 'Hardware is in use.'");
+						case AudioConverterError.NoHardwarePermission:
+							throw new InvalidOperationException ("ComponentList could not be set: 'No Hardware permission.'");
+						case AudioConverterError.None:
+							return;
+						default:
+							throw new InvalidOperationException ("ComponentList could not be set.");
```
Comment (rolfbjarne) [inline src/AudioUnit/AudioComponent.cs]: Include the error number in the exception message.

File: src/AudioUnit/AudioUnit.cs [hunk 15]
```diff
@@ -352,6 +356,50 @@ public enum AudioUnitClumpID // UInt32 in AudioUnitParameterInfo
...
+	public enum AUParameterEventType : uint
+	{
+		Immediate = 1,
+		Ramped = 2,
+	}
+
+	[StructLayout (LayoutKind.Sequential)]
+	public struct AudioUnitParameterEvent
+	{
+		public uint scope;
```
Comment (rolfbjarne) [inline src/AudioUnit/AudioUnit.cs]: Public members should be properly cased, not lower-cased (this goes for all the public fields in this structure and its inner structures).

File: src/AudioUnit/AudioUnit.cs [hunk 16]
```diff
@@ -1032,6 +1085,10 @@ class AURenderCallbackStrct {
 		static extern AudioUnitStatus AudioUnitSetParameter (IntPtr inUnit, AudioUnitParameterType inID, AudioUnitScopeType inScope,
 			uint inElement, float inValue, uint inBufferOffsetInFrames);
 
+		[iOS (2,0)]
+		[DllImport (Constants.AudioUnitLibrary)]
+		static extern  AudioUnitStatus AudioUnitScheduleParameters (IntPtr inUnit, AudioUnitParameterEvent inParameterEvent, uint inNumParamEvents);
```
Comment (rolfbjarne) [inline src/AudioUnit/AudioUnit.cs]: minor: two spaces between `extern` and `AudioUnitStatus`.

File: src/AudioUnit/AudioUnit.cs (hunk 6 above)
Comment (spouliot) [inline src/AudioUnit/AudioUnit.cs]: Likely Hight Quality, so `HrtfHQ` (value of `2` + `HQ`) or `Hrtfhq`, would be correct .net names.

> And we now have a new rendering algorithm in this Spatial Mixer, called HRTFHQ, high quality.

See transcript of https://developer.apple.com/videos/play/wwdc2017/501/

**Comment by spouliot** [comment]
@mandel-macaque PR failures must be reviewed and mapped to existing (or new) bugs. A link here to the bug, a link in the bug to the PR.

File: tests/monotouch-test/AudioToolbox/AudioComponentTest.cs [hunk 17]
```diff
@@ -0,0 +1,59 @@
...
+	
+	[TestFixture]
+	[Preserve (AllMembers = true)]
+	public class AudioComponentTest {
+
+		[Test]
+		public void GetSetComponentList ()
+		{
+			if (IntPtr.Size == 4)
+				Assert.Ignore ("API only available in 64 bits.");
```
Comment (rolfbjarne) [inline tests/monotouch-test/AudioToolbox/AudioComponentTest.cs]: The API isn't 32-bit only, it's iOS 11 (and equivalent versions for other platforms) only.

This means you should use `TestRuntime.AssertXodeVersion (9, 0)` instead.

File: src/AudioUnit/AudioUnit.cs [hunk 18]
```diff
@@ -1032,6 +1085,10 @@ class AURenderCallbackStrct {
 		static extern AudioUnitStatus AudioUnitSetParameter (IntPtr inUnit, AudioUnitParameterType inID, AudioUnitScopeType inScope,
 			uint inElement, float inValue, uint inBufferOffsetInFrames);
 
+		[iOS (2,0)]
```
Comment (spouliot) [inline src/AudioUnit/AudioUnit.cs]: ^ remove, undecorated API are assumed to be part of our minimum support (so start with 6.x).

File: src/AudioUnit/AudioComponent.cs [hunk 19]
```diff
@@ -38,6 +38,161 @@
...
+		public ResourceUsageInfo ResourceUsage { 
+			get {
+				return GetStrongDictionary<ResourceUsageInfo> (resourceUsageK);
+			} 
+			set {
+				SetNativeValue (resourceUsageK, value.Dictionary, true);
+			}
+		}
+
+		public NSString[] Tags {
```
Comment (spouliot) [inline src/AudioUnit/AudioComponent.cs]: are those tags constants ? (e.g. from `[Field]`)
if not it looks like they should be exposed as `string` (not `NSString`).

File: src/AudioUnit/AudioComponent.cs [hunk 20]
```diff
@@ -224,6 +379,58 @@ public XamCore.AppKit.NSImage GetIcon ()
 			return new XamCore.AppKit.NSImage (AudioComponentGetIcon (handle));
 		}
 #endif
+		[NoWatch, NoTV, Mac (10,13), iOS (11,0)]
+		[DllImport (Constants.AudioUnitLibrary, EntryPoint = "AudioUnitExtensionSetComponentList")]
```
Comment (spouliot) [inline src/AudioUnit/AudioComponent.cs]: it's not public so there's no need to have an `EntryPoint`, so name the method with it's native name

File: src/AudioUnit/AudioComponent.cs [hunk 21]
```diff
@@ -224,6 +379,58 @@ public XamCore.AppKit.NSImage GetIcon ()
...
+		static extern int /* OSStatus */ SetComponentList (IntPtr /* CFString */ extensionIdentifier, /* CFArrayRef */ IntPtr audioComponentInfo);
+
+		[NoWatch, NoTV, Mac (10,13), iOS (11,0)]
+		[DllImport (Constants.AudioUnitLibrary, EntryPoint = "AudioUnitExtensionCopyComponentList")]
```
Comment (spouliot) [inline src/AudioUnit/AudioComponent.cs]: same

File: src/AudioUnit/AudioComponent.cs [hunk 22]
```diff
@@ -224,6 +379,58 @@ public XamCore.AppKit.NSImage GetIcon ()
...
+						var result = (AudioConverterError) SetComponentList (cfString.Handle, array.Handle);
+						switch (result) {
+						case AudioConverterError.HardwareInUse:
+							throw new InvalidOperationException ("ComponentList could not be set: 'Hardware is in use.'");
+						case AudioConverterError.NoHardwarePermission:
+							throw new InvalidOperationException ("ComponentList could not be set: 'No Hardware permission.'");
+						case AudioConverterError.None:
+							return;
+						default:
+							throw new InvalidOperationException ($"ComponentList could not be set, error {result.ToString ()}");
```
Comment (spouliot) [inline src/AudioUnit/AudioComponent.cs]: not sure there's much value in all the similar (but different) strings, they just end up taking 2xspace in the metadata (unicode).
`throw new InvalidOperationException (result.ToString ());` 
should give the required information back to the developer (the stack trace will point to `ComponentList`).

File: src/audiounit.cs [hunk 23]
```diff
@@ -437,7 +486,7 @@ interface AUParameter : NSSecureCoding
 		[Export ("setValue:originator:atHostTime:eventType:")]
 		void SetValue (float value, IntPtr originator, ulong hostTime, AUParameterAutomationEventType eventType);
 
-		[iOS (10,0), Mac (10,12, onlyOn64 : true)]
+		[iOS (10,0), Mac (10,12, onlyOn64 : true), Watch (4, 0), TV (11, 0)]
 		[TV (10,0)]
```
Comment (spouliot) [inline src/audiounit.cs]: it's decorated with both `[TV(10,0]` and `[TV(11,0)]` ?
please check back if Xcode8 has this for tvOS

Comment (mandel-macaque) [inline src/audiounit.cs]: Me culpa, I missed the other line of attrs.

File: src/AudioUnit/AudioComponent.cs (hunk 22 above)
Comment (mandel-macaque) [inline src/AudioUnit/AudioComponent.cs]: Ok, we can remove that , it is the main two errors we will get, but lets do that and the string should be helpfull enough when searching the web.

File: src/AudioUnit/AudioComponent.cs (hunk 19 above)
Comment (mandel-macaque) [inline src/AudioUnit/AudioComponent.cs]: Those are not constants, are given in the plist. I'll set it to work in c# strings.

File: src/AudioUnit/AudioUnit.cs [hunk 24]
```diff
@@ -352,6 +356,50 @@ public enum AudioUnitClumpID // UInt32 in AudioUnitParameterInfo
...
+			}
+
+
+			[FieldOffset (0)]
+			public RampStruct Ramp;
+
+			[StructLayout (LayoutKind.Sequential)]
+			public struct ImmediateStruct
+			{
+				public uint bBufferOffset;
```
Comment (rolfbjarne) [inline src/AudioUnit/AudioUnit.cs]: Typo: `BufferOffset` instead of `bBufferOffset`.

File: src/AudioUnit/AudioComponent.cs [hunk 25]
```diff
@@ -38,6 +38,162 @@
...
+				SetBooleanValue (sandboxSafeK, value);
+			}
+		}
+
+		public ResourceUsageInfo ResourceUsage { 
+			get {
+				return GetStrongDictionary<ResourceUsageInfo> (resourceUsageK);
+			} 
+			set {
+				SetNativeValue (resourceUsageK, value.Dictionary, true);
```
Comment (spouliot) [inline src/AudioUnit/AudioComponent.cs]: `value.Dictionary` -> `NullReferenceException`
if `null` is allowed then you should use `value?.Dictionary`
if not then throw an exception

File: src/AudioUnit/AudioComponent.cs [hunk 26]
```diff
@@ -38,6 +38,162 @@
...
+		static NSString exceptionK = new NSString ("temporary-exception.files.all.read-write");
+
+		public ResourceUsageInfo () : base () {}
+
+		public ResourceUsageInfo (NSDictionary dic) : base (dic) {}
+
+		public string[] IOKitUserClient { 
+			get {
+				var array = GetNativeValue<NSArray> (userClientK);
+				return NSArray.StringArrayFromHandle (array.ClassHandle);
```
Comment (spouliot) [inline src/AudioUnit/AudioComponent.cs]: That should be `.Handle` not `.ClassHandle`
Please complete unit tests so it confirms the value you set can be read back (to the same strings).

File: src/AudioUnit/AudioComponent.cs [hunk 27]
```diff
@@ -38,6 +38,162 @@
...
+			} 
+			set {
+				SetArrayValue (userClientK, value);
+			}
+		}
+
+		public string[] MachLookUpGlobalName { 
+			get {
+				var array = GetNativeValue<NSArray> (globalNameK);
+				return NSArray.StringArrayFromHandle (array.ClassHandle);
```
Comment (spouliot) [inline src/AudioUnit/AudioComponent.cs]: same

File: src/AudioUnit/AudioComponent.cs [hunk 28]
```diff
@@ -38,6 +38,162 @@
...
+			} 
+			set {
+				SetNativeValue (resourceUsageK, value.Dictionary, true);
+			}
+		}
+
+		public string[] Tags { 
+			get {
+				var array = GetNativeValue<NSArray> (tagsK);
+				return NSArray.StringArrayFromHandle (array.ClassHandle);
```
Comment (spouliot) [inline src/AudioUnit/AudioComponent.cs]: `.Handle` + additional tests

File: src/AudioUnit/AudioComponent.cs [hunk 29]
```diff
@@ -224,6 +395,54 @@ public XamCore.AppKit.NSImage GetIcon ()
...
+						for (var i = 0; i < result.Length; i++) {
+							result [i] = new AudioComponentInfo (dics[i]);
+						}
+						return result;
+					}
+				}
+			}
+			set {
+				using (var cfString = new CFString (Name)) {
+					var dics = new NSDictionary [value.Length];
```
Comment (spouliot) [inline src/AudioUnit/AudioComponent.cs]: that's a `NRE` if value is `null`
check if `null` is valid (or not) and fix / add validations

Comment (spouliot) [inline src/AudioUnit/AudioComponent.cs]: also add a unit test for `null`
